utils/custodian/custodian.go

```diff
@@ -0,0 +1,128 @@
+package custodian
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/brave-intl/bat-go/utils/altcurrency"
+	appctx "github.com/brave-intl/bat-go/utils/context"
+	loggingutils "github.com/brave-intl/bat-go/utils/logging"
+	"github.com/google/uuid"
+	"github.com/shopspring/decimal"
+)
+
+// TransactionStatus - stringer repr of transaction status
+type TransactionStatus fmt.Stringer
+
+// Transaction - interface defining what a transaction is
+type Transaction interface {
+	GetIdempotencyKey(context.Context) (fmt.Stringer, error)
+	GetAmount(context.Context) (decimal.Decimal, error)
+	GetCurrency(context.Context) (altcurrency.AltCurrency, error)
+	GetDestination(context.Context) (fmt.Stringer, error)
+	GetSource(context.Context) (fmt.Stringer, error)
+}
+
+// implementation of Transaction
+type transaction struct {
+	IdempotencyKey *uuid.UUID              `json:"idempotencyKey,omitempty"`
+	Amount         decimal.Decimal         `json:"amount,omitempty"`
+	Currency       altcurrency.AltCurrency `json:"currency,omitempty"`
+	Destination    *uuid.UUID              `json:"destination,omitempty"`
+	Source         *uuid.UUID              `json:"source,omitempty"`
+}
+
+// GetItempotencyKey - implement transaction
+func (t *transaction) GetIdempotencyKey(context.Context) (fmt.Stringer, error) {
+	return t.IdempotencyKey, nil
+}
+
+// GetDestination - implement transaction
+func (t *transaction) GetDestination(context.Context) (fmt.Stringer, error) {
+	return t.Destination, nil
+}
+
+// GetSource - implement transaction
+func (t *transaction) GetSource(context.Context) (fmt.Stringer, error) {
+	return t.Source, nil
+}
+
+// GetAmount - implement transaction
+func (t *transaction) GetAmount(context.Context) (decimal.Decimal, error) {
+	return t.Amount, nil
+}
+
+// GetCurrency - implement transaction
+func (t *transaction) GetCurrency(context.Context) (altcurrency.AltCurrency, error) {
+	return t.Currency, nil
+}
+
+// NewTransaction - create a new transaction
+func NewTransaction(ctx context.Context, idempotencyKey, destination, source *uuid.UUID, currency altcurrency.AltCurrency, amount decimal.Decimal) (Transaction, error) {
+	return &transaction{
+		IdempotencyKey: idempotencyKey,
+		Destination:    destination,
+		Source:         source,
+		Currency:       currency,
+		Amount:         amount,
+	}, nil
+}
+
+// Custodian - interface defining what a custodian is
+type Custodian interface {
+	SubmitTransactions(context.Context, ...Transaction) error
```

Comment:
I think the current Bitflyer implementation, these return the same signature. It looks like here we are changing that. Do we want to have `SubmitTransaction` also return `(map[string]TransactionStatus, error)` ? We would need to pull some of the grouping logic out of the `SubmitTransactions` into a pre-step, but then could consolidate some logic. Thoughts?